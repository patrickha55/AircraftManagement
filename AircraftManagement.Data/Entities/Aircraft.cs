using System.ComponentModel.DataAnnotations;

namespace AircraftManagement.Data.Entities
{
    public class Aircraft
    {
        [Required]
        [Range(minimum: 1, maximum: int.MaxValue)]
        public int Id { get; set; }
        [Required]
        [StringLength(maximumLength: 100, MinimumLength = 2)]
        public string Model { get; set; }
        [Required]
        [StringLength(maximumLength: 100, MinimumLength = 2)]
        public string Manufacturer { get; set; }
        public string Image { get; set; }

        public int RoleId { get; set; }
        public Role Role { get; set; }
    }
}
