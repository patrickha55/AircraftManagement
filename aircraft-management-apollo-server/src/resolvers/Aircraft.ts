const Aircraft = {
    role: async ({ roleId }: any, _: any, { roleApi }: any) => {
        try {
            const response = await roleApi.getRoleAsync(roleId);

            return response.data;
        } catch (error: any) {
            if (error.response) {
                console.log(error.response);
            } else if (error.request) {
                console.log(error.request);
            } else {
                console.log(error);
            }
        }
    }
};

export default Aircraft;