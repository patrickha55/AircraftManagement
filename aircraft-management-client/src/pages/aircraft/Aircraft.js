import { gql, useMutation, useQuery } from '@apollo/client';
import DeleteIcon from '@mui/icons-material/Delete';
import { Button, IconButton } from '@mui/material';
import { useEffect, useState } from 'react';
import CreateAircraftModal from '../../components/createAircraftModal/CreateAircraftModal';
import React from 'react';

export const GET_AIRCRAFTS = gql`
    query Aircrafts{
        aircrafts {
            id
            manufacturer
            model
            image
            role {
                name
            }
        }
    }
`;

export const DELETE_AIRCRAFT = gql`
    mutation ($id: Int!) {
        deleteAircraft (id: $id)
    }
`;

const Aircraft = () => {
    const [aircrafts, setAircrafts] = useState([]);

    const { data, error, loading: loadingAircraft } = useQuery(GET_AIRCRAFTS);
    const [deleteAircraft, {
        loading,
        error: mutationError
    }
    ] = useMutation(DELETE_AIRCRAFT);

    const [open, setOpen] = useState(false);

    useEffect(() => {
        if (data) {
            setAircrafts(data.aircrafts);
        }
    }, [data]);


    const handleModal = () => setOpen(prev => !prev);

    const createAircraftHandler = (aircraft) => {

        setAircrafts([...aircrafts, aircraft]);

    };

    const deleteHandler = async (id) => {
        const result = await deleteAircraft({
            variables: {
                id: id
            }
        });

        if (result.data.deleteAircraft) {
            setAircrafts(aircrafts.filter(p => p.id !== id));
        };
    };

    if (loadingAircraft) return <p>Loading...</p>
    if (error) return <h2 className='alert alert-danger'>{error.message}</h2>

    return (
        <div className='container'>
            <h1 className='text-center my-5'>Aircraft</h1>
            <Button variant="contained" color="primary" onClick={handleModal}>
                Create Aircraft
            </Button>
            <CreateAircraftModal
                open={open}
                handleModal={handleModal}
                onCreateAircraft={createAircraftHandler} />
            {
                aircrafts.map((a) =>
                    <div key={a.id} className="card w-75 mx-auto mt-4 shadow-lg" >
                        <div className="card-body">
                            <img src={a.image} className="card-img-top  border-2 shadow-lg" alt="..."></img>
                            <h2>{a.manufacturer}</h2>
                            <div className='row'>
                                <div className="col-6">
                                    <h5 className="card-text">Model: </h5>
                                </div>
                                <div className="col-6">
                                    <h5 className="card-text text-secondary">{a.model}</h5>
                                </div>
                                <div className="col-6">
                                    <h5 className="card-text">Role: </h5>
                                </div>
                                <div className="col-6">
                                    <h5 className="card-text text-secondary">{a.role.name}</h5>
                                </div>
                                <div className="col-6">
                                    Update
                                </div>
                                <div className="col-6">
                                    <IconButton aria-label="delete"
                                        onClick={async () => { await deleteHandler(a.id) }}
                                        disabled={loading && true}
                                    >
                                        <DeleteIcon data-testid={`delete${a.model}`} />
                                    </IconButton>
                                </div>
                                <div className="col-12">
                                    {
                                        mutationError && <small className='text-danger'>Something went wrong. Please try again!</small>
                                    }
                                </div>
                            </div>
                        </div>
                    </div>
                )
            }
        </div>
    )
}

export default Aircraft;