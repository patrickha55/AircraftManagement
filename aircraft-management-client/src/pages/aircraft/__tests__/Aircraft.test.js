import React from 'react';
import { ReactDOM } from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { render, fireEvent, cleanup, screen } from '@testing-library/react';
import Aircraft, { GET_AIRCRAFTS } from '../Aircraft';
import { MockedProvider } from '@apollo/react-testing';
import renderer from 'react-test-renderer';
import waait from 'waait';


afterEach(cleanup);

const aircrafts = [
    { id: 1, manufacturer: "Boeing", model: "A380", image: "A380Image", roleId: 1, role: { id: 1, name: "Wide body aircraft" } },
    { id: 2, manufacturer: "Lockheed Martin", model: "F22", image: "F22Image", roleId: 3, role: { id: 3, name: "Air superiority fighter" } }
];

const aircraftMocks = [
    {
        request: {
            query: GET_AIRCRAFTS,
        },
        result: {
            data: {
                aircraft: aircrafts
            }
        }
    }
];

it("renders without error.", () => {
    render(
        <MockedProvider mocks={aircraftMocks} addTypename={false} >
            <Aircraft />
        </MockedProvider>
    );
});

it('should renders loading state initially', () => {
    //arrange
    const component = renderer.create(
        <MockedProvider mocks={[]}>
            <Aircraft />
        </MockedProvider>,
    );

    const tree = component.toJSON();

    // assert
    expect(tree).toMatchSnapshot();

    expect(tree.children).toContain('Loading...');
});

// it('should renders aircrafts', async () => {
//     //arrange
//     render(
//         <MockedProvider mocks={aircraftMocks} addTypename={false} >
//             <Aircraft />
//         </MockedProvider>
//     );

//     await waait(2000);
//     //act
//     // jest.mock("aircrafts", () => aircrafts);

//     // const h2s = queryAllByText(/Boeing|Lockheed Martin/i);
//     //assert

//     screen.debug();
// });
