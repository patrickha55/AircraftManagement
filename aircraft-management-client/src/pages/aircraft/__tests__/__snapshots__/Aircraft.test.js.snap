// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should renders loading state initially 1`] = `
<p>
  Loading...
</p>
`;
